xgcm/grid.py

```diff
@@ -4,1199 +4,57 @@
+from .axis import Axis
+from .grid_ufunc import (
+    GridUFunc,
+    _check_data_input,
+    _GridUFuncSignature,
+    _has_chunked_core_dims,
+    _maybe_unpack_vector_component,
+    _reattach_coords,
+    apply_as_grid_ufunc,
+)
+from .metrics import iterate_axis_combinations
+from .padding import pad
 
-        if not keep_coords:
-            warnings.warn(
-                "The keep_coords keyword argument is being deprecated - in future it will be removed "
-                "entirely, and the behaviour will always be that currently given by keep_coords=True.",
-                category=DeprecationWarning,
-            )
+try:
+    import numba  # type: ignore
 
-        position_from, old_dim = self._get_position_name(da)
-        try:
-            new_dim = self.coords[position_to]
-        except KeyError:
-            raise KeyError("Position '%s' was not found in axis.coords." % position_to)
-
-        orig_dims = da.dims
-
-        coords = OrderedDict()
-        dims = []
-        for d in orig_dims:
-            if d == old_dim:
-                dims.append(new_dim)
-                # only add coordinate if it actually exists
-                # otherwise this creates a new coordinate where before there
-                # was none
-                if new_dim in self._ds.coords:
-                    coords[new_dim] = self._ds.coords[new_dim]
-            else:
-                dims.append(d)
-                # only add coordinate if it actually exists...
-                if d in da.coords:
-                    coords[d] = da.coords[d]
-
-        # add compatible coords
-        if keep_coords:
-            for c in da.coords:
-                if c not in coords and set(da[c].dims).issubset(dims):
-                    coords[c] = da[c]
-
-        return xr.DataArray(data_new, dims=dims, coords=coords)
-
-    def _get_position_name(self, da):
-        """Return the position and name of the axis coordinate in a DataArray."""
-        for position, coord_name in self.coords.items():
-            # TODO: should we have more careful checking of alignment here?
-            if coord_name in da.dims:
-                return position, coord_name
-
-        raise KeyError(
-            "None of the DataArray's dims %s were found in axis "
-            "coords." % repr(da.dims)
-        )
+    from .transform import transform
+except ImportError:
+    numba = None
 
-    def _get_axis_dim_num(self, da):
-        """Return the dimension number of the axis coordinate in a DataArray."""
-        _, coord_name = self._get_position_name(da)
-        return da.get_axis_num(coord_name)
+# Only need this until python 3.8
+try:
+    from typing import Literal
+except ImportError:
+    from typing_extensions import Literal  # type: ignore
 
 
-_XGCM_BOUNDARY_KWARG_TO_XARRAY_PAD_KWARG = {
-    "periodic": "wrap",
-    "fill": "constant",
-    "extend": "edge",
-}
+def _maybe_promote_str_to_list(a):
+    # TODO: improve this
```

Comment:
How do we need to improve this?

Reply:
I don't remember... `_maybe_promote_str_to_list` is only used within `set_metrics`, and seems conceptually similar to our `_map_kwargs_over_axes`, so maybe could be refactored to use that. This is v low priority though.